Kawaiwan add execution metrics

charleskwwan opened the pull request:
# Description
- Added metrics to execution framework
- Added tests for execution framework, including those that do not involve metrics
# Testing done
- Created and ran unit tests for metrics and rest of execution framework
- Ran itests from examples
# Note to reviewers
- Metrics code is contingent on yelp_meteorite being present. If not, a mock object is returned so that calls to counters or timers do not break.
- Changes do not yet include a way to optionally install yelp_meteorite (or mock)  as the dependency_links option in setuptools was deprecated. An extra requirements file for Yelp dependencies might be required instead, as Paasta does it.

jolynch commented on task_processing/metrics.py:
```diff
@@ -0,0 +1,54 @@
+except Exception:
+    METRICS_ENABLED = False
+
+TASKPROC_TASK_LAUNCHED_COUNT = 'taskproc_task_launched_count'
```
This is overly verbose, I don't think you need to namespace everything by taskproc.

charleskwwan replied:
Was trying to follow yelp_kew's example, and yelp_meteorite's best practices. Thoughts?

jolynch replied:
Oh, the metric name (RHS) is fine imo, the LHS is redundant (don't need TASKPROC in the taskproc package)

charleskwwan replied:
Ah, agreed.

charleskwwan replied:
Sure, I'll change the RHS.

jolynch replied:
@sagar8192 sure the metric name itself can have it (rhs), but I'm saying the LHS is redundant. We're in the taskproc package, we don't need constants to be prefixed with taskproc IMO.

jolynch commented on tests/unit/plugins/mesos/mesos_executor_test.py:
```diff
@@ -43,7 +47,10 @@ def test_creates_execution_framework_and_driver(mesos_executor):
     )
 
-    mesos_executor.driver.run.assert_called_with()
+    threading.Thread.assert_called_once_with(
```
don't use assert_called_once_with, instead check that call_args == some list

https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html

charleskwwan replied:
Agreed.

jolynch commented on task_processing/metrics.py:
```diff
@@ -0,0 +1,54 @@
+TASKPROC_TASK_KILLED_COUNT = 'taskproc_task_killed_count'
+TASKPROC_TASK_LOST_COUNT = 'taskproc_task_lost_count'
+TASKPROC_TASK_ERROR_COUNT = 'taskproc_task_error_count'
+TASKPROC_TASK_MAX_RETRIES_COUNT = 'taskproc_task_max_retries_count'
```
Can we encapsulate all these constants into the mesos executor? I don't feel like these are general metrics.

charleskwwan replied:
Agreed.

sagar8192 replied:
+1

jolynch commented on setup.py:
```diff
@@ -33,5 +33,6 @@
     extras_require={
         # We can add the mesos specific dependencies here
         'mesos_executor': ['pymesos'],
+        'metrics': ['yelp-meteorite', 'mock']
```
having mock as a runtime dependency is a pretty strange, can we just return an actual function?

charleskwwan replied:
We are returning a mock object in metrics if yelp_meteorite isn't around, so that functions called on counters/timers won't break. Thought about trying to create a class with the combined protocol of y_m's counters/timers/gauges, but it might fall short. Thoughts?

jolynch replied:
let's avoid a runtime dependency on mock if we can. We can emulate whichever portion of the yelp-meteorite API we use :-)

charleskwwan replied:
Alright. I'll hash one out and get it to review.

jolynch commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -264,11 +308,39 @@ def create_new_docker_task(self, offer, task_config, available_ports):
+            'framework_role': self.role
+        }
+
+        create_counter(TASKPROC_TASK_LAUNCHED_COUNT, default_dimensions)
```
I feel like we can DRY this up

charleskwwan replied:
Agreed.

jolynch commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -8,6 +8,23 @@
 from pyrsistent import PRecord
 from six.moves.queue import Queue
 
+from task_processing.metrics import create_counter
```
Let's DRY this up. Maybe put the metrics in a collection of some sort?

charleskwwan replied:
Agreed.

charleskwwan commented:
Also to reviewers, please also review the tests. Because its a large new file, the diff is not loaded by default. You'll need to do so manually.

sagar8192 commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,571 @@
 @pytest.fixture
 def ef():
-    return ExecutionFramework("name", "role")
+    threading.Thread = mock.Mock()
```
I don't this is gonna work. You need a fixture to mock out threading.Thread. You can pass it to ef().

charleskwwan replied:
It does seem to work for me. pdb shows that the Thread method called in ExecutionFramework is a mock object. But I suppose using a fixture makes it cleaner?

sagar8192 commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,571 @@
+    agent_id = 'fake_agent_id'
+    # Need to save and restore original time function for test output
+    original_time_func = time.time
+    time.time = mock.Mock(return_value=2.0)
```
You need to create a fixture for time.time

charleskwwan replied:
Agreed.

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -65,6 +82,9 @@ def __init__(
         self.task_metadata = {}
 
+        self._initialize_metrics()
+        self._offer_time = None
```
Agreed.

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -202,6 +228,17 @@ def get_tasks_to_launch(self, offer):
+                )
+                get_metric(TASKPROC_TASK_QUEUED_TIME_TIMER).record(
+                    (time.time() -
+                     self.task_metadata[task.task_id].time_launched) * 1000,
```
I was trying to follow yelp_meteorite's standard that time sent be in milliseconds. I did notice that kew does it seconds however. I suppose this comment is for our own consistency?

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -264,11 +308,39 @@ def create_new_docker_task(self, offer, task_config, available_ports):
+        }
+
+        create_counter(TASKPROC_TASK_LAUNCHED_COUNT, default_dimensions)
+        create_counter(TASKPROC_TASK_FINISHED_COUNT, default_dimensions)
```
I can do that.

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -264,11 +308,39 @@ def create_new_docker_task(self, offer, task_config, available_ports):
+        create_counter(TASKPROC_TASK_KILLED_COUNT, default_dimensions)
+        create_counter(TASKPROC_TASK_LOST_COUNT, default_dimensions)
+        create_counter(TASKPROC_TASK_ERROR_COUNT, default_dimensions)
+        create_counter(TASKPROC_TASK_MAX_RETRIES_COUNT, default_dimensions)
```
Agreed.

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -294,6 +366,13 @@ def resourceOffers(self, driver, offers):
             self.driver = driver
 
+        current_offer_time = time.time()
+        if self._offer_time:
```
Agreed.

charleskwwan commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -132,12 +152,18 @@ def enqueue_task(self, task_config):
 
+        get_metric(TASKPROC_TASK_QUEUED_COUNT).count(
+            1,
+            {'task_id': task_config.task_id}
```
Agreed.

charleskwwan commented on task_processing/metrics.py:
```diff
@@ -0,0 +1,54 @@
+import mock
```
I think @jolynch's idea that we move the mesos-specific constants to the executor is good, so we don't need to move metric itself.

charleskwwan commented:
### New changes based on comments, please review
- Fixed metric names
- Converted assert_called_once_with to call_args
- DRY for metric initialization
- Replaced mock with DummyMetricType in metrics
- Renamed offer_time to last_offer_time
- Removed extra dimensions in metric calls
- Kept time records in seconds not ms
- Added metrics for blacklisted agents and stuck tasks
- Removed requeued and retry specific metrics
- Add fixture for time.time and threading.Thread in ef tests

@sagar8192 @jolynch

sagar8192 commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -365,14 +432,17 @@ def statusUpdate(self, driver, update):
             self.task_metadata.pop(task_id, None)
+            get_metric(TASK_FINISHED_COUNT).count(1)
+
+        unsuccessful_states = {
```
This can be moved to the constructor.

charleskwwan replied:
Agreed.

sagar8192 commented on task_processing/metrics.py:
```diff
@@ -0,0 +1,47 @@
+    if METRICS_ENABLED:
+        return _registered_metrics.get(name)
+    else:
+        class DummyMetricType(object):
```
Can we initialize this class in the module, create a global object and return the same object every time?

charleskwwan replied:
Sure.

sagar8192 commented on tests/unit/plugins/mesos/mesos_executor_test.py:
```diff
@@ -8,6 +11,7 @@
 def mesos_executor(mocker, request):
     mocker.patch.object(me_module, 'ExecutionFramework')
     mocker.patch.object(me_module, 'MesosSchedulerDriver')
+    threading.Thread = mock.Mock()
```
We need mock patch here.

charleskwwan replied:
Agreed.

sagar8192 commented on task_processing/plugins/mesos/execution_framework.py:
```diff
@@ -264,11 +298,34 @@ def create_new_docker_task(self, offer, task_config, available_ports):
         self.stopping = True
 
+    def _initialize_metrics(self):
+        default_dimensions = {
```
Can you add a TODO here for adding mesos cluster dimension?

charleskwwan replied:
Sure.

keymone commented:
your diff includes irrelevant commits, probably due to merges, could you rebase please?

charleskwwan commented on requirements-dev.txt:
```diff
@@ -1,4 +1,6 @@
+mock
```
Right, forgot.

charleskwwan commented on tests/unit/plugins/mesos/mesos_executor_test.py:
```diff
@@ -1,13 +1,23 @@
+def mesos_executor(mocker, request, mock_Thread):
     mocker.patch.object(me_module, 'ExecutionFramework')
     mocker.patch.object(me_module, 'MesosSchedulerDriver')
+    threading.Thread = mock.Mock()
```
Thought I removed this. Agreed.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    # Avoid sleeping during test run and piggyback to stop infinite loop
+    def stop_killing(task_id):
+        ef.stopping = True
+    time.sleep = mock.Mock(side_effect=stop_killing)
```
Agreed.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+        task_state_ts=1.0
+    )
+    ef.create_new_docker_task = mock.Mock(return_value='fake_docker_task')
+    mock_docker_task = ef.create_new_docker_task.return_value
```
I need this to ensure the rest of tasks_to_launch executes entirely.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    tasks_to_launch = ef.get_tasks_to_launch(fake_offer)
+
+    assert len(tasks_to_launch) == 0
+    assert not ef.task_queue.empty()
```
Agreed.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+        available_ports
+    )
+
+    assert ef.task_metadata[task_id].agent_id == 'fake_agent_id'
```
Agreed.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    ef.task_metadata = ef.task_metadata.set(task_id, task_metadata)
+    ef.resourceOffers(ef.driver, [fake_offer])
+
+    assert ef.driver.launchTasks.call_count == 1
```
Sure.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+        ef.offer_decline_filter
+    )
+    assert fake_driver.suppressOffers.call_count == 1
+    assert ef.are_offers_suppressed
```
Sure.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    assert fake_driver.declineOffer.call_args == mock.call(
+        fake_offer.id,
+        ef.offer_decline_filter
+    )
```
Sure.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    assert fake_driver.declineOffer.call_args == mock.call(
+        fake_offer.id,
+        ef.offer_decline_filter
+    )
```
Sure.

charleskwwan commented on tests/unit/plugins/mesos/execution_framework_test.py:
```diff
@@ -1,54 +1,559 @@
+    assert fake_driver.declineOffer.call_args == mock.call(
+        fake_offer.id,
+        ef.offer_decline_filter
+    )
```
Sure.

sagar8192 commented on Makefile:
```diff
@@ -2,6 +2,18 @@
 
+YELP_IDX="https://pypi.yelpcorp.com/simple"
+ORIG_DEPS=mesos_executor
+EXTRA_DEPS=mesos_executor,metrics
```
rm?

charleskwwan replied:
Oops, forgot. Agreed.

sagar8192 commented on tests/unit/plugins/mesos/mesos_executor_test.py:
```diff
@@ -43,7 +52,10 @@ def test_creates_execution_framework_and_driver(mesos_executor):
     )
 
-    mesos_executor.driver.run.assert_called_with()
+    threading.Thread.assert_called_once_with(
```
mock_thread

charleskwwan replied:
Right.
